### ISSUE TITLE
not resolving host if it comes from google ip

### ISSUE THREAD

**Issue by scratchmex** [damomurf/coredns-tailscale#67]
I know this sounds strange but believe me, sometimes it resolves and sometimes it doesn't:

```
2024-01-06T04:40:28.028 app[xyz] gdl [info] [INFO] 211.25.206.147:53467 - 18628 "A IN somehost.ipn.example.com. udp 49 true 4000" NOERROR qr,aa 74 0.00005624s
                                                                                                                                  ^^^^^^^
2024-01-06T05:11:49.815 app[xyz] gdl [info] [INFO] 204.194.238.66:45387 - 9330 "A IN somehost.ipn.example.com. udp 49 true 1410" NOERROR qr,aa,cd 74 0.00008114s 
                                                                                                                                 ^^^^^^^
2024-01-06T04:40:28.057 app[xyz] gdl [info] [INFO] 109.0.64.17:36563 - 56454 "A IN somehost.ipn.example.com. udp 49 false 1232" NOERROR qr,aa 74 0.00010105s
                                                                                                                                ^^^^^^^
2024-01-06T04:40:28.068 app[xyz] gdl [info] [INFO] 172.253.229.196:47130 - 39090 "A IN somehost.ipn.example.com. udp 49 true 1400" NXDOMAIN qr,aa,cd 38 0.00003451s 
                                                                                                                                   ^^^^^^^^
```

as you can see the last query is returning NXDOMAIN and that ip is from google. I don't know what it have to do with google but believe me as I am doing

```
dig @208.67.222.222 somehost.ipn.example.com. +short
100.111.222.333
$ dig @1.1.1.1 somehost.ipn.example.com. +short
100.111.222.333
$ dig @8.8.8.8 somehost.ipn.example.com. +short
<empty>
```

I discover this when doing a dns check in one of those online pages which do many dns query around the globe. only the google (and the ones fallbacking to it) where failing for no reason

coredns config
```
{$NS_ZONE} {
  records {
    @ 60 IN A 1.2.3.4
    # apparently we need the NS records here too
    @ 60 IN NS ns1
    @ 60 IN NS ns2
    # some resolvers will recheck the entries for DNS glue records at the delegate nameserver.
    # Manually specify these hosts, since they won't appear in the Tailscale node list.
    ns1 60 IN A {$NS_IP}
    ns2 60 IN A {$NS_IP}
  }

  tailscale {$NS_ZONE} {
    authkey {$TS_AUTHKEY}
    fallthrough
  }

  log
  errors
}
```

**Comment by scratchmex** [comment]
@willnorris do you have any clue why this might be? I am tagging you because I am trying to setup the tailscale dns resolver from your blog post. I don't know if you have encountered this already

**Comment by willnorris** [comment]
Yes, I have consistently had issue with Google DNS wanting to resolve my custom DNS:

```
% dig @1.1.1.1 samson.ipn.willnorris.net +short
100.111.203.54

% dig @8.8.8.8 samson.ipn.willnorris.net +short
<empty>
```

I've not tracked down what the problem is, but have actually stopped using Google DNS as upstream for my local network for exactly this reason.

**Comment by scratchmex** [comment]
the strangest part is that the coredns service is responding NXDOMAIN to google dns but not to others. what? how does it know? in the logs (above) I see the response

**Comment by damomurf** [comment]
@scratchmex I’d presume this is an upstream CoreDNS issue.

I’m curious what the use case is for having these addresses “on the internet”? I find the Tailscale split-dns functionality works a treat here and I forward the domain in question in that config to a container running coredns-tailscale that only ever serves requests for that domain?

**Comment by scratchmex** [comment]
I actually don't necessarily need the ips public. I actually was trying to make coredns listen on the tailscale interface but I can't. Whenever I do `dig @<coredns-ts-ip>` it doesn't respond. Do you have tips for a good setup? The objective is to have a custom magicdns domain like `ipn.yourdomain.tld`

**Comment by damomurf** [comment]
@scratchmex Ok, so I found a similar issue trying to the same thing after @willnorris’s changes to implement tsnet in the plugin. I noticed his Dockerfile is doing the following, which I hadn’t tried so I’ll see if that helps in my case.
```
RUN setcap cap_net_bind_service=+ep 
```

**Comment by scratchmex** [comment]
I have that too. I suspect is a fly.io problem

**Comment by damomurf** [comment]
@scratchmex I replicated the listener never receiving dns requests issue locally and it’s occurred to me that there’s no way tsnet can override the underlying CoreDNS listener and it doesn’t even try to.

I wonder if @willnorris has successfully used this and could share the required configuration if so?

**Comment by willnorris** [comment]
wow, fascinating!  I had no idea... good find @underfake !

Now that I know what to look for, this was apparently discussed pretty publicly at the time:
 - https://groups.google.com/g/public-dns-announce/c/MLsrx8dI2n4
 - https://www.theregister.com/2023/01/19/google_dns_queries/
 - and many others: https://www.google.com/search?q=google+dns+random+casing
